src/core/geometry/qgsgeos.cpp

```diff
@@ -548,7 +587,21 @@ double QgsGeos::distance( const QgsAbstractGeometry *geom, QString *errorMsg ) c
     return distance;
   }
 
-  geos::unique_ptr otherGeosGeom( asGeos( geom, mPrecision ) );
+  // GEOSPreparedDistance_r is not able to properly compute the distance if one
+  // of the geometries if a vertical line (LineString Z((X Y Z1, X Y Z2, ..., X Y Zn))).
+  // In that case, replace `geom` by a single point.
+  // However, GEOSDistance_r works.
+  std::unique_ptr< QgsAbstractGeometry > preparedGeom;
+  if ( mGeosPrepared && isZVerticalLine( geom ) )
+  {
+    preparedGeom.reset( geom->vertexAt( QgsVertexId( 0, 0, 0 ) ).clone() );
+  }
+  else
+  {
+    preparedGeom.reset( geom->clone() );
```

Comment:
This is inefficient -- it's forcing an unnecessary clone in the standard (non vertical line) case. 

Move `geos::unique_ptr otherGeosGeom'` to before this check, and avoid setting preparedGeom when we don't need it.

Reply:
Done.